### ISSUE TITLE
Improve engine handling of PyObjects

### ISSUE THREAD

**Issue by jjbrosnan** [deephaven/deephaven-core#4166]
**Description**

PyObjects in the query language can be tricky to deal with. Specifically, if you have a PyObject column containing lists or arrays, converting them to a better type (double[], String[], or any Java array) just doesn't work at all. You have to use a Python function to do the conversion:

```python
import pandas as pd

mylist = [["a", "b", "C", "DEF"], ["g", "h", "i", "k"]]

df = pd.DataFrame({"Col1": mylist})

from deephaven.pandas import to_table
from deephaven import dtypes as dht

def pyobj_to_arr(item):
    return dht.array(dht.string, item)

t = to_table(df).update(["Col1 = (String[])pyobj_to_arr(Col1)"])
```

It would be nice if this conversion could happen without the Python function:

```python
import pandas as pd

mylist = [["a", "b", "C", "DEF"], ["g", "h", "i", "k"]]

df = pd.DataFrame({"Col1": mylist})

t = to_table(df).update(["Col1 = (String[])Col1"])
```

**Comment by jmao-denver** [comment]
The solution to the problem is 
```
import pandas as pd

mylist = [["a", "b", "C", "DEF"], ["g", "h", "i", "k"]]

df = pd.DataFrame({"Col1": mylist})

from deephaven.pandas import to_table
from deephaven import dtypes as dht

def pyobj_to_arr(item):
    return dht.array(dht.string, item)

t = to_table(df).update(["Col1 = (String[])pyobj_to_arr(Col1)"])
```

**Comment by chipkent** [comment]
I think there should be a few supported paths and questions:

1) For type hints, I would expect this to work:
```
def func(x: int) -> Sequence[int]:
    return [0, x, x+1]

t2 = t1.update("Y = func(X)")
```

or alternatively:
```
def func(x: int) -> Sequence[np.int]:
    return [0, x, x+1]

t2 = t1.update("Y = func(X)")
```

2) Explicit casts could be useful but may be redundant with #1.  For example, this could be a clean way to transparently inject the `dht.array(dht.long,...)` cast:
```
def func(x: int):
    return [0, x, x+1]

t2 = t1.update("Y = (int[])func(X)")
```

3) For high performance, numba arrays should also be supported.  I don't know the synaxt off hand for doing arrays in numba, but I think it would be something like:
```
@vectorize([double[:](int64)]
def func(x):
    return [0, x, x+1]

t2 = t1.update("Y = func(X)")
```

4) I think numpy should also work, but it would be a subset of #1 and #2 syntactically.  With numpy, there are a few other considerations:
* It should avoid data copies, since the backing array should transfer between java and numpy.
* Numpy appears to have new typehints for array type.  These should work when type hints are used.  https://numpy.org/devdocs/reference/typing.html#numpy.typing.NDArray.
* When casts are used, we need to decide if the cast fails or converts when the types do not exactly match.
* Numpy arrays can be multidimensional.  I suppose that we should only support 1D out of the box and fail on any higher dimensions.
* If a numpy type hint does not contain an element type, do we fail?

5) There is an alternative idea that comes with different performance tradeoffs.  Instead of dealing with primitive arrays, results could be wrapped as Deephaven Arrays.  I might suggest not doing this, but I could probably be convinced either way.

6) At some point, we may also want to support Set/Dict.  Some thought should be given to the design and how it would play well or poorly in that scenario. 

7) If the user explicitly creates their own Java array, do we have a nice way to express the problem without having to cast?
```
def pyobj_to_arr(item: str) -> JAVA_ARRAY_TYPE:
    return dht.array(dht.string, item)

t = to_table(df).update(["Col1 = pyobj_to_arr(Col1)"])

**Comment by rcaudy** [comment]
It's a little ugly that we have `cast` operators that aren't casts, but rather translations to a new data type. This is more of a C++ behavior than a Java behavior, and I think it runs the risk of confusing users.

**Comment by rcaudy** [comment]
I think your point (5) is probably not a good, idea, @chipkent. We should convert to flat, native Java arrays; a vector might be nice for deferred conversions from Python types to Java types, but the cost of accessing a vector wrapping a Python list or similar would be terrible over JNI.

**Comment by rcaudy** [comment]
For your points (6), @chipkent , don't we already have exposed tools for creating Java maps, etc? Implicit conversion seems expensive.

**Comment by chipkent** [comment]
I'm reviewing this ticket now that we have merged #4579.  I would expect this PR to resolve most or all of the items in this ticket.

... but the example (1) above is buggy.  See #4631.
```python
from deephaven import empty_table
from typing import Sequence
import numpy as np

def func(x: int) -> Sequence[np.int32]:
    return [0, x, x+1]

t2 = empty_table(10).update("Y = func(i)")
```

Additionally, the initial description in this ticket is pathological and relies on Pandas column-type support.

```python
import numpy as np
import pandas as pd
from deephaven.pandas import to_table, to_pandas

# Pandas seems to have no clue about vector columns, so it just classifies them as `object` type.
# the DH conversion then can't infer anything better.  Not sure we can improve on this.
mylist = [["a", "b", "C", "DEF"], ["g", "h", "i", "k"]]
df = pd.DataFrame({"Col1": mylist})
print(df.dtypes)
print(to_pandas(to_table(df).meta_table))

# Just throwing values in a DF column also results in an untyped pandas column.
# I would classify this case as user error.
mylist = ["a", "b", "c"]
df = pd.DataFrame({"Col1": mylist})
print(df.dtypes)
print(to_pandas(to_table(df).meta_table))

# If the user specifies a column type in Pandas, they can get a column type that DH handles correctly.
mylist = pd.Series(["a", "b", "c"], dtype="string")
df = pd.DataFrame({"Col1": mylist})
print(df.dtypes)
print(to_pandas(to_table(df).meta_table))

# Pandas column typing is very dumb.  It can't figure out the type here.
mylist = np.array(["a", "b", "c"], dtype=np.str_)
df = pd.DataFrame({"Col1": mylist})
print(df.dtypes)
print(to_pandas(to_table(df).meta_table))
```
<img width="483" alt="image" src="https://github.com/deephaven/deephaven-core/assets/5250374/9332fd00-f50c-42cf-8a96-95d06d9c82e1">

**Comment by jmao-denver** [comment]
With #4647 in, the original code can be rewritten as:
```
import pandas as pd

mylist = [["a", "b", "C", "DEF"], ["g", "h", "i", "k"]]

df = pd.DataFrame({"Col1": mylist})

from deephaven.pandas import to_table
from deephaven import dtypes as dht

def pyobj_to_arr(item) -> List[str]:
    return item

t = to_table(df).update(["Col1 = pyobj_to_arr(Col1)"])
```
Even though it seems not quite as straightforward as improving to_table() to auto convert the 'list' type column into Java arrays, it is actually quite difficult to get such auto-conversion done correctly and reliably. Since this is probably not a very common use case and a workaround is available, I agree that we should close it instead of spending more energy on further improvement.
